File: 2023/day11/Python/part1.py
from itertools import combinations


inp = open('../input.txt', 'r').read()
space = inp.splitlines()

row_len = inp.index('\n')
blank_row = '.' * row_len

blank_col_indices = []
expanded_space = []
galaxy_coords = []


# after every blank row, add in another blank row
for line in space:
    expanded_space.append(line)

    if line == blank_row:
        expanded_space.append(blank_row)

# get the index of every blank column
for i in range(row_len):
    is_col_blank = True

    for line in expanded_space:
        if line[i] == '#':
            is_col_blank = False
            break

    if not is_col_blank:
        continue
    else:
        blank_col_indices.append(i)


# after every blank column, add in another blank column
for offset, idx in enumerate(blank_col_indices):
    for j, line in enumerate(expanded_space):
        expanded_space[j] = line[:idx + offset] + '.' + line[idx + offset:]


space = [list(line) for line in expanded_space]

for y, line in enumerate(expanded_space):
    for x, char in enumerate(line):
        if char == '#':
            galaxy_coords.append( (y, x) )


galaxy_pairs = combinations(galaxy_coords, 2)
sum = 0

for pair in galaxy_pairs:
    # |y2 - y1| + |x2 - x1|
    shortest_path = abs(pair[1][0] - pair[0][0]) + abs(pair[1][1] - pair[0][1])

    sum += shortest_path


print(sum)

File: 2023/day11/C/part1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


int main()
{
    FILE* input = fopen("../input.txt", "r");

    int row_len = 0;
    while ( getc(input) != '\n' )
        row_len++;

    char line[row_len + 2]; // +2 for '\n' and '\0'
    char blank_row[row_len];

    // make a blank row
    memset(blank_row, '.', row_len);
    blank_row[row_len] = '\0';

    int blank_row_indices[row_len];
    int blank_col_indices[row_len];
    int num_blank_rows = 0;
    int num_blank_cols = 0;

    int galaxies = 0;


    // get the index of every blank row
    for (int i = 0; fgets(line, sizeof(line), input) != NULL; i++)
    {
        line[row_len] = '\0';
        if (strcmp(line, blank_row) == 0)
        {
            blank_row_indices[num_blank_rows++] = i;
        }
    }
    rewind(input);


    // get the index of every blank column
    int is_col_blank = 1;
    for (int i = 0; i < row_len; i++)
    {
        is_col_blank = 1;
        while (fgets(line, sizeof(line), input) != NULL)
        {
            if (line[i] == '#')
            {
                is_col_blank = 0;
                galaxies++;
            }
        }

        rewind(input);

        if (!is_col_blank)
            continue;
        else
            blank_col_indices[num_blank_cols++] = i;
    }


    int y_coords[galaxies];
    int x_coords[galaxies];
    int cnt_y = 0, cnt_x = 0;
    int j = 0;

    // For every galaxy, we will offset that galaxy's (y, x) coordinate with the
    // number of blank rows & columns before it, times the expansion rate.
    for (int y = 0; fgets(line, sizeof(line), input) != NULL; y++)
    for (int x = 0; x < row_len; x++)
    {
        if (line[x] == '#')
        {
            cnt_y = cnt_x = 0;

            // count the number of blank rows before this galaxy
            for (int i = 0; i < num_blank_rows; i++)
            {
                if (y > blank_row_indices[i])
                    cnt_y++;
            }

            // count the number of blank columns before this galaxy
            for (int i = 0; i < num_blank_cols; i++)
            {
                if (x > blank_col_indices[i])
                    cnt_x++;
            }

            // coords after the expansion
            y_coords[j  ] = y + cnt_y;
            x_coords[j++] = x + cnt_x;
        }
    }



    fclose(input);
    long sum = 0;

    // for every galaxy pair, calculate the shortest path between them
    for (    j =   0; j < galaxies; j++)
    for (int k = j+1; k < galaxies; k++)
    {
        // |y2 - y1| + |x2 - x1|
        sum += abs(y_coords[k] - y_coords[j]) + abs(x_coords[k] - x_coords[j]);
    }


    printf("%ld\n", sum);

    return EXIT_SUCCESS;
}